controllers/postgresqldatabase_controller.go

```diff
@@ -223,14 +263,90 @@ func (r *PostgreSQLDatabaseReconciler) EnsurePostgreSQLDatabase(log logr.Logger,
 			log.Error(err, "failed to close database connection", "host", host, "database", "postgres", "user", credentials.Name)
 		}
 	}()
-	err = postgres.Database(log, db, host, postgres.Credentials{
-		Name:     name,
-		User:     user,
-		Password: password,
-		Shared:   isShared,
-	})
+	err = postgres.Database(log, db, host, params.Target)
 	if err != nil {
-		return fmt.Errorf("create database %s on host %s: %w", name, connectionString, err)
+		return fmt.Errorf("create database %s on host %s: %w", params.Target.Name, connectionString, err)
 	}
 	return nil
 }
+
+// credentials returns the correct `postgres.Credentials` based on the values
+// of `params.Host` and `params.HostCredentials` fields of `params`. Exactly
+// one of these fields must be set, and if `Host` is set, then this method
+// will return the credentials from `r.HostCredentials` map, otherwise it will
+// search the Kubernetes namespace for a `PostgreSQLHostCredentials` with the
+// name specified in `params.HostCredentials`.
+func (r *PostgreSQLDatabaseReconciler) credentials(ctx context.Context, params *EnsureParams) (string, *postgres.Credentials, error) {
+	// If the `HostCredentials` field is populated and the `Host` field is
+	// empty, then return credentials stored in the corresponding
+	// `PostgreSQLHostCredentials` resource.
+	if params.Host == "" && params.HostCredentials != "" {
+		// Fetch the `PostgreSQLHostCredentials` from the API.
```

Comment:
For readability can we extract the body of this `if` into a new method maybe named `remoteCredentials`? Is it becomes "if hostCredentials is set -> return r.remoteCredentials()` in pseudo code? That will help us avoid this very long branching method and instead make it a decisition method that knows "what" instead of "how".

Does that make sense?